Add `gen` blocks and reserve keyword in Rust 2024 (#3513)
rust-lang/rfcs: text/0000-gen-fn.md

writes values to mutable state. With this RFC, one can write the `for` loop
and still get a lazy iterator of values.

As an example, here are multiple ways to write an iterator over something that contains integers,
only keep the odd integers, and multiply all of them by 2:

```rust
// `Iterator` methods
fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
    values.filter(|value| value.is_odd()).map(|value| value * 2)
}

// `struct` and manual `impl`
fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
    struct Foo<T>(T);
    impl<T: Iterator<Item = u32>> Iterator<Item = u32> for Foo<T> {
        type Item = u32;
        fn next(&mut self) -> Option<u32> {
            loop {
                let value = self.0.next()?;
                if value.is_odd() {
                    return Some(x * 2)
                }
            }
        }
    }
    Foo(values)
}

// `gen block`
fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
    #[rustc_gen] {
        for value in values {
            if value.is_odd() {
                yield value * 2;
            }
        }
    }
}

// `gen fn`
#[rustc_gen]

> fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {

Author: It makes the initial implementation harder, as the hir would have to extract a type from this arbitrary ast node. Thus I'd prefer not doing this as part of the very first impl PR

> fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {

Reviewer: It's also worth pointing out that at the point that we're writing `Item = u32` in an attribute, we would have to ask ourselves why we're not just writing `-> impl Iterator<Item = u32` somewhere.

> fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {

Reviewer: Given Oli's response, I'm going to mark this conversation as resolved.  Thanks @estebank  for pointing this out as it does highlight interesting other design directions.  If there's more to say, please unresolve it of course.

    for value in values {
        if value.is_odd() {
            yield value * 2;
        }
    }
}
```

Iterators created with `#[rustc_gen]` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).
See [the unresolved questions][#unresolved-questions] for whether `#[rustc_gen]` iterators are fused or may behave strangely after having returned `None` once.
Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## New keyword

Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, ... named `gen` must be renamed.

## Returning/finishing an iterator

> `#[rustc_gen]` block's trailing expression must be of the unit type or the block must diverge before reaching its end.

Reviewer 2: Is "a block's trailing expression must be of the unit type" sufficient to imply that uses of `return` within the block must either be given `()` or no expression at all?

I personally suspect not, since I can imagine linguistic constructs that would do some kind of coercion of the unit value from the trailing expression while still allowing `return` to take a non-unit expression as input.

And therefore, based on that suspicion, I recommend this bit of text be revised to explicitly say (somehow) that `return` within the block is restricted to unit. (E.g. " uses of `return` within the block must either be given `()` or no expression at all.")

Author: My proposal is to treat `return` and finishing the block exactly the same and require them both to be returning unit

Reviewer: We've now revised this section to be more clear and comprehensive with respect to what is allowed in the trailing position and as an argument to return.  Thanks for pointing out the ambiguities here.

(I'm going to mark this conversation as resolved, but of course, if you have any observations about the new language, please open one for that.)

### Diverging iterators

For example, a `#[rustc_gen]` block that produces the infinite sequence `0, 1, 0, 1, 0, 1, ...`, will never return `None`
from `next`, and only drop its captured data when the iterator is dropped.

```rust
#[rustc_gen] {
    loop {
        yield 0;
        yield 1;
    }
}
```

If a `#[rustc_gen]` block panics, the behavior is very similar to `return`, except that `next` unwinds instead of returning `None`.

## Error handling

Within `#[rustc_gen]` blocks, the `?` operator desugars differently from how it desugars outside of `#[rustc_gen]` blocks.

> Instead of returning the `Err` variant, `foo?` yields the `Err` variant and then `return`s immediately afterwards.

Reviewer 2: ```suggestion
If `foo` evaluates to the `Err` variant, then instead of returning it, the desugaring of `foo?` now yields the `Err` variant and then `return`s immediately afterwards.
```

(I'm torn about this suggestion, to be honest. My motivation for suggesting it is that I thought it was not 100% clear that you were strictly discussing the case where `foo` itself returns `Err` here.)

Also: I assume the interaction with `From::from` is the same as it is for `?` outside of `#[rustc_gen]` blocks?

Reviewer 2: (Oh, maybe this is all covered in the Reference-level explanation...)

Reviewer: We've now revised this section to be more clear and comprehensive.  Thanks for pointing out the ambiguities here.

This creates an iterator with `Iterator::Item`'s type being `Result<T, E>`.

> Once a `Some(Err(e))` is produced via `?`, the iterator returns `None` on the subsequent call to `Iterator::next`.

Reviewer 2: ```suggestion
Once a `Some(Err(e))` is returned from `Iterator::next` via a use of `?` within the `#[rustc_gen]`, the iterator returns `None` on the subsequent call to `Iterator::next`.
```

Reviewer 2: Question: if one *intends* to return an `Some(Err(e))` but allow subsequent calls to `next()` to return a non-`None` result, then the appropriate way to encode that (instead of the `?`-operator) is to use `yield Err(e)`, right?

Author: Yes

Reviewer: We've now revised this section in a way that should encompass the suggestion.  Based on that and oli's answer, I'll mark this as resolved.  But of course, if there's more to say here, please unresolve it.

In contrast to other code where you can use `?`, `#[rustc_gen]` blocks do not need to have a trailing `Ok(x)` or `x` expression.
Returning from a `#[rustc_gen]` block will make the `Iterator` return `None`, which needs no value.
Instead, all `yield` operations must be given a `Result`.

The `?` operator on `Option`s will `yield None` if it is `None`, and require passing an `Option` to all `yield` operations.
As an example:

```rust
let x = some_option?;
yield Some(x + 1)
```

> will yield `None` if `some_option` is `None`, but `Some(x + 1)` otherwise.

Author: The desugaring is the easiest part to change before stabilization. And it won't be part of the very first PR anyway

repeatedly produce values of the unit type.

# Drawbacks
[drawbacks]: #drawbacks

It's another language feature for something that can already be written entirely in user code.

In contrast to `Generator`, `#[rustc_gen]` blocks that produce `Iterator`s cannot hold references across `yield` points.
See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.

[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives
## Keyword

We could use `iter` as the keyword.
I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.
The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.
I believe connecting the `Iterator` trait with `iter` blocks is the right choice,
but that would require us to carve out many exceptions for this keyword as `iter` is used for module names and method names everywhere (including libstd/libcore).
It may not be much worse than `gen` (see also [the unresolved questions][#unresolved-questions]).

> We may want to use `gen` for full on generators in the future.

Reviewer 2: If we were to use `gen` for full-on generators in the future, would they look something like `gen |arg1, arg2, ...| { ... }` ?

I.e. in that world, would the iterator-oriented feature in this RFC "continue" to be written `gen { ... }` (which I assume is the future expected state of `#[rustc_gen] { ... }`, and one would be distinguishing that from full-on generators soley via the presence/absence of the parameter list `|arg1, arg2, ...|` ?

(I'm not commenting yet on whether that mode of distinguishing is good or bad. I just want to make sure I understand all this first.)

Author: Yea, full on generators/coroutines need new syntax, but the `gen` keyword may also be reused in that syntax

Reviewer 3: I'll try my best to answer this.

I don't think it can ever be enough to distinguish between the two kinds of generators just based on a parameter lists: after all this RFC already introduces "generator functions" with input parameters. Which means it's not unthinkable we'd want to enable "generator closures" to work in the future too.

For clarity I'll start referring to "full-on generators" as "co-routines" going forward instead. Co-routines are a lot more complicated than the current iterator is. They cannot be mapped to the existing `Iterator` trait, and instead require [a new trait](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) which can take arguments into the yield statement. As well as distinguish between a "yield type" and a "return type". Put together this would probably need to look something like this:

```rust
// example trait definition
pub trait CoRoutine<R = ()> {
    type Yield;
    type Return;
    fn resume(&mut Self, resume: R) -> CoRoutineState<Self::Yield, Self::Return>;
}
pub enum CoRoutineState <Y, R> {
    Yielded(Y),
    Complete(R),
}

// syntax example (entirely made up, not an actual proposal)
co(Food) fn eat_food(cat: &mut Cat) -> Nap yields Purr {
// ^                                   ^          ^
// |                                   |          |
// |                                   |  The type yielded (output)
// |                                   |
// |                           The return type (output)
// |
// The type returned from `yield` (input)
    for 0..10 {
        // We yield a `Purr`, but `yield` returns `Food`.
        let food = yield Purr::new();
        cat.eat(food);
    }
    Nap::new() // We end the function by returning a `Nap`
}
```

If you compare this with the `gen fn` feature proposed by this RFC, you'll notice that "co-routines" contain a lot more moving parts which need to be dfeined. And even if both features are conceptually related, they are actually quite different - and we should probably think of them as separate. Does that make sense?

> ## Do not do this

Reviewer 4: I have used `iter::from_fn` for simple cases myself. It's definitely the closest we have to generators at the moment.

> ## Do not do this

Reviewer: We've added an example of this and discussion of it to the document.  Thanks for pointing this out.

One alternative is to keep adding more helper methods to `Iterator`.
It is already hard for new Rustaceans to be aware of all the capabilities of `Iterator`.
Some of these new methods would need to be very generic.
While it's not an `Iterator` example, [`array::try_map`][] is something that has very complex diagnostics that are hard to improve, even if it's nice once it works.

Users can use crates like [`genawaiter`](https://crates.io/crates/genawaiter) or [`propane`](https://crates.io/crates/propane) instead.
`genawaiter` works on stable and provides `gen!` macro blocks that behave like `#[rustc_gen]` blocks, but don't have compiler support for nice diagnostics or language support for the `?` operator. The `propane` crate uses the `Generator` trait from nightly and works mostly
like `#[rustc_gen]` would.

[`array::try_map`]: https://doc.rust-lang.org/std/primitive.array.html#method.try_map

> ## `return` statements `yield` one last element

Reviewer 2: I find it funny that none of the concrete examples in this subsection actually use explicit `return` statements.

Is there a potential world where we would continue to treat a trailing `()` expression as terminating the iteration with no associated yield, while an *explicit* `return expr` would yield the value of `expr` and then terminate the expression? Would that world address any/all of the concerns raised below about trailing expressions? (Would that "benefit" justify the damage wrought by now introducing a distinction between trailing expressions vs `return` that we had not previously had?)

Author: I don't think that distinction would help, as then you'd rule out early abort via `return`, unless you special case `return ()` which gets really weird if your iterator also yields `()`

Reviewer 5: If you have an explicit return point where you want to yield one final element that can be written `return yield foo;` (given my assumption of how the desugaring will look).

Reviewer 2: > as then you'd rule out early abort via return

oh, maybe I wasn't clear. By "and then terminate the expression", I *meant* to still early abort. Yield the returned value and don't iterate furhter.

Reviewer: We've made a note of this in the relevant section of the RFC.

```rust
#[rustc_gen] fn foo() -> i32 {
    42; // note the semicolon
}
```

would then not return a value.

Furthermore this would make it unclear what the behaviour of

```rust
#[rustc_gen] fn foo() {}
```

is supposed to be, as it could be either `std::iter::once(())` or `std::iter::empty::<()>()`

## different syntax for `#[rustc_gen] fn`:

This RFC explicitly picks an attribute, as that has no conflicts with any other syntax, even within macros, and
does not pick any option that may influence how experimental users think about syntax.
There are many options to choose from, and we'll have to decide on one before stabilization.
Some options are:

```rust
fn foo(args) yield item
fn foo(args) yields item
fn foo(args) => item
fn* foo(args) -> item // or any of the `fn foo` variants for the item type
gen fn foo(args) // or any of the above variants for the item type
gen foo(args) // or any of the above variants for the item type
generator fn foo(args) // or any of the above variants for the item type
```

The design space here is very large, but either way, I propose to reserve the `gen` keyword.

# Prior art

> [prior-art]: #prior-art

Reviewer: Thanks.  We've added this example to the document.

[unresolved-questions]: #unresolved-questions

## Keyword

Should we use `iter` as the keyword, as we're producing `Iterator`s?
We could use `gen` as proposed in this RFC and later extend its abilities to more powerful generators.

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=efeacb803158c2ebd57d43b4e606c0b5)

```rust
#![feature(generators)]
#![feature(iter_from_generator)]

fn main() {
    let mut it = std::iter::from_generator(|| {
        yield 1
    });

    assert_eq!(it.next(), Some(1));
    assert_eq!(it.next(), None);
    it.next(); // panics
}
```

## Panicking

What happens when `Iterator::next` is called again on a `#[rustc_gen]` block that panicked? Do we need to poison the iterator?

## Contextual keyword

Popular crates (like `rand`) have methods called [`gen`][Rng::gen]. If we forbid those, we are forcing those crates to make a major version bump when they update their edition, and we are requiring any users of those crates to use `r#gen` instead of `gen` when calling that method.

We could choose to use a contextual keyword and only forbid `gen` in

* bindings,
* field names (due to destructuring bindings),
* enum variants,

> * and type names

Author: I didn't see any in the parser when I checked

This should avoid any parsing issues around `gen` followed by `{` in expressions.

[Rng::gen]: https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen

## `Iterator::size_hint`

Should we try to compute a conservative `size_hint`? This will reveal information from the body of a generator,

> but at least for simple cases users will likely expect `size_hint` to not just be the default.

Reviewer 6: A manual mechanism to specify `size_hint` can be added later, but also it would be possible to automatically calculate it in rustc if every looped iterator has a `size_hint` set: single `yield` statements increment the counter, every `for _ in iter` adds `iter.size_hint()` to it, and nested iterators are multiplied. This breaks down as soon as there's a single iterator without a valid `size_hint` though, which is why the manual mechanism should also exist. All of this is *not* needed for an MCP though.

There are infinite options to choose from if we want such a feature, so I'm listing general ideas:

### Do nothing, just use loops

```rust
for x in iter {
    yield x
}
```

### Language support

we could do something like postfix `yield` or an entirely new keyword, or...

```rust
iter.yield
```

### stdlib macro

We could add a macro to the standard library and prelude.
The macro would expand to a `for` loop + `yield`.

```rust
yield_all!(iter)
```

## Complete `Generator` support

We already have a `Generator` trait on nightly that is more powerful than the `Iterator`
API could possibly be.

1. it uses `Pin<&mut Self>`, allowing self-references in the generator across yield points
2. it has arguments (`yield` returns the arguments passed to it in the subsequent invocations)

Similar to the ideas around `async` closures,
I think we could argue for `Generators` to be `#[rustc_gen]` closures while `#[rustc_gen]` blocks are a simpler concept that has no arguments and only captures variables.

Either way, support for full `Generator`s should be discussed and implemented separately,

> as there are many more open questions around them beyond a simpler way to write `Iterator`s.

Reviewer 6: We should explicitly mention the possibility of writing

```
gen {
    let x = yield 0;
    println!("{x:}");
}
```
which is supported by `Generator::resume(val)`, but isn't by `Iterator` (today? we could add a new method `Iterator::resume` that by default calls `Iterator::next`).

Author: That behaviour needs a more closure like syntax (like nightly generators) anyway, as you also need the argument before the first yield. Even more syntax will make it easier to distinguish between generators and coroutines or anything in between

Author: I'll add a note to the full generators section

```rust
match foo.branch() {
    ControlFlow::Break(err) => {
        yield R::from_residual(err);
        return;
    },
    ControlFlow::Continue(val) => val,
}
```

This is the same behaviour that `collect::<Result<_, _>>()` performs
on iterators over `Result`s.

## Implementation

This feature is mostly implemented via existing generators.
We'll need additional desugarings and lots of work to get good diagnostics.

### `#[rustc_gen] fn`

`#[rustc_gen] fn` desugars to the function itself with the return type replaced by `impl Iterator<Item = $ret>` and its body wrapped in a `#[rustc_gen]` block.
A `#[rustc_gen] fn`'s "return type" is its iterator's `yield` type.

A `#[rustc_gen] fn` captures all lifetimes and #[rustc_gen]eric parameters into the `impl Iterator` return type (just like `async fn`).
If more control over captures is needed, type alias impl trait can be used when it is stabilized.

Like other uses of `impl Trait`, auto traits are revealed without being specified.

### `#[rustc_gen]` blocks

`#[rustc_gen]` blocks are the same as an unstable generator...

* ...without arguments,

> * ...with an additional check forbidding holding borrows across `yield` points,

Reviewer 5: Is there any appetite to consider this point in type-space between `Iterator` and `Generator`, where internal borrows are supported? It seems very limiting from my previous experiments to not be able to use `for x in &vec` at all.

Author: yes, imo that's what `Generator` should be, and our `Generator` could be renamed to `Coroutine` or sth. Figuring out lending iterators and self-referential iterators is something desirable, but with this RFC I'm trying to solve writing `Iterator`s and not closing the door on any of the future designs we'll want in this space. Once we have an incremental small step, doing the next one is easier than doing multiple at once.

Reviewer 2: >  It seems very limiting from my previous experiments to not be able to use `for x in &vec` at all

Can someone clarify for me what this limitation is referring to? In what context the `for x in &vec` not usable? 

Are you e.g. saying that the `&Vec` is a parameter to the `#[rustc_gen] fn` and the `x` is being yielded, like so:

```rust
#[rustc_gen] fn demo<'a, X>(vec: &'a Vec<X>) -> &'a X {
     for x in vec { yield x; }
}
```

or do you mean that the `Vec` is a parameter to the `#[rustc_gen] fn` (and the `x` is still being yielded), like so:

```rust
#[rustc_gen] fn demo<X>(vec: Vec<X>) -> &'??? X {
    let borrowed = &vec;
    for x in borrowed { yield x; }
}
```

(The latter is problematic, at least at a surface level, because we do not have existential lifetimes. But I want to first figure out which case you two are talking about before I attempt to dissect it further.)

Reviewer 5: The second case, but not attempting to yield a borrow, just use it internally to compute an owned value that gets yielded.

Reviewer 7: (NOT A CONTRIBUTION)

There's two different issues:

1. A generator can't *store* references to its own state unless its pinned, but the `Iterator` interface doesn't take a pinned reference.
2. A generator can't *yield* references to its won state unless its interface borrows from itself, but the `Iterator` interface doesn't have the "lending" interface.

(The second issue also has the problem of having no clear way to express the lifetime, as you noted @pnkfelix.)

This RFC punts both issues, though probably at least the 1st issue needs to be figured out before you can stabilize something.

Author: From the implementation side the first issue is resolved by checking for `Unpin` just like `std::iter::from_generator` does. This will require additional work for good diagnostics, but the raw logic for it exists.

Reviewer 5: That check is not necessary since the `#[rustc_gen]` transform is in control of the generator transform which is applied, it just needs to not set the flag that allows self-borrows (the difference between `|| yield;` and `static || yield;` generators). That has decent enough diagnostics already, they may just need the exact text changed depending on what this feature is called in documentation:

```
error[E0626]: borrow may still be in use when generator yields
6 |         for x in &values { yield *x; }
  |                  ^^^^^^^   -------- possible yield occurs here
```

The issue for this RFC is more: is there anything that might be designed differently to be forward-compatible with a syntax that allows self-borrows and generates a future `PinnedIterator` trait (or however it's spelled)?

- Feature Name: `gen-fn`
- Start Date: 2023-10-10
- RFC PR: [rust-lang/rfcs#3513](https://github.com/rust-lang/rfcs/pull/3513)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Reserve the `gen` keyword in the 2024 edition.

Add `gen {}` blocks to the language. These implement `Iterator` by `yield`ing
elements. This is simpler and more intuitive than creating a custom type and
manually implementing `Iterator` for that type, which requires writing an
explicit `Iterator::next` method body. This is a change similar to adding `async
{}` blocks that implement `Future` instead of having to manually write futures
and their state machines.

> Furthermore, add `gen fn` to the language. `gen fn foo(arg: X) -> Y` desugars to

Reviewer 6: Given that there's more finality in the choice of syntax now than what was discussed earlier, I am concerned about this syntax, particularly using the return type to mean the yielded type. This is a one way door change that precludes extending `gen fn` to support non-() return types in the future. Given that `yield` is already a reserved keyword, I would prefer to use that instead of `->` (or use `yields` as a contextual keyword).

Reviewer: It may make sense to just remove `gen fn` as part of the normative proposal in this document.  This doesn't need to be decided to reserve the keyword or even for an initial stabilization of an expression syntax.

There's also the issue that there has been some discussion around the idea that we may have erred on this point with `async fn`, i.e. maybe it should have included the `-> impl Future<..>` after all.  Eliding that has made a number of problems more difficult as bounds on the returned `Future` cannot be expressed.  It's something we should think about here also.

Anyway, we're planning some deeper revisions, and we'll definitely think about this point.

Reviewer 8: (Apologies if this has been covered elsewhere in the discussion, I'm behind on the comments.)

I don't know if removing `gen fn` from part of the syntax helps much from a keyword standpoint because we'll still presumably want something like `gen { }` blocks, and that needs a keyword.

As far as syntax goes, right now I'm kind of fond of `gen fn foo() yield i32 { }`, or even just `fn foo() yield i32 { }`.

One reason I like this is that it generalizes pretty well to coroutines (at least for the ones where `resume` doesn't take an argument), since we could write `fn foo() -> bool yield i32 { }` for a coroutine that returns `bool` and yields some number of `i32`s along the way.

Reviewer 9: I would like to see `gen fn` removed from the normative proposal as well, for now, just so that we can go ahead and approve the `gen` keyword and `gen`/`async gen` blocks.

Reviewer: This was removed.

    std::iter::from_fn(move || {
        loop {
            let value = values.next()?;
            if value % 2 == 1 {
                return Some(value * 2);
            }
        }
    })
}

// `struct` and manual `impl`
fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
    struct Foo<T>(T);
    impl<T: Iterator<Item = u32>> Iterator<Item = u32> for Foo<T> {
        type Item = u32;
        fn next(&mut self) -> Option<u32> {
            loop {
                let value = self.0.next()?;
                if value.is_odd() {
                    return Some(x * 2)
                }
            }
        }
    }
    Foo(values)
}

// `gen block`
fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
    gen {
        for value in values {
            if value.is_odd() {
                yield value * 2;
            }
        }
    }
}

// `gen fn`
gen fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {
    for value in values {
        if value.is_odd() {
            yield value * 2;
        }
    }
}
```

Iterators created with `gen` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).

> See [the unresolved questions][unresolved-questions] for whether `gen` iterators are fused or may behave strangely after having returned `None` once.

Reviewer 10: The unresolved questions section doesn't mention this at all?

Reviewer: The document has since changed to settle the question of whether generators are fused (they are), and whether they might later implement `FusedIterator` is now listed as a future possibility.

Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## New keyword

> Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed.

Reviewer 10: ```suggestion
Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed or migrated to raw identifier syntax (`r#gen`).
```

Some crates may be unable to rename items in their public API.

Reviewer: We've since updated the document to cover this.  Thanks for the note.

argument returns a value indicating "do not short circuit"
(e.g. `Option::Some(..)`, `Result::Ok(..)`, `ControlFlow::Continue(..)`), that
value becomes the result of the expression as usual.  When its argument
returns a value indicating that short-circuiting is desired
(e.g. `Option::None`, `Result::Err(..)`, `ControlFlow::Break(..)`), the value
is first yielded (after being converted by `From::from` as usual), then the
block returns immediately.

Even when `?` is used within a `gen` block, the block must return a
value of type unit or `!`.  That is, it does not return a value of `Some(..)`,
`Ok(..)`, or `Continue(..)` as other such blocks might.

However, note that when `?` is used within a `gen` block, all `yield`
statements will need to be given an argument of a compatible type.  For
example, if `None?` is used in an expression, then all `yield` statements will
need to be given arguments of type `Option`.

## Fusing

Like `Generators`, `Iterator`s produced by `gen` panic when invoked again after they have returned `None` once.
This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially

> produce infinite values of the unit type.

Reviewer 10: As the RFC no longer aims to be "experimental," this should probably just state the desired eventual behavior?

Reviewer: This section was updated.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation
## New keyword

In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `k#gen` to get the same features.

## Error handling

`foo?` in `gen` blocks will stop iteration after the first error by desugaring to:

```rust
match foo.branch() {
    ControlFlow::Break(err) => {
        yield R::from_residual(err);
        return;
    },
    ControlFlow::Continue(val) => val,
}
```

This is the same behaviour that `collect::<Result<_, _>>()` performs

> on iterators over `Result`s.

Reviewer 11: I don't get this comparison, and I find the `?` extremely weird. Could you elaborate on the example?

`collect` turns an iterator of `Result` into a single `Result<Vec<_>>`, using the first error. How does that have anything to do with what generators do under the proposed semantics?

Reviewer 12: The similarity is that both stop at the first error. But agreed that there's a difference in the types used; with `?` as described here you'd still have an iterator over a Result.

Reviewer 11: Okay, so saying it's the "same" behavior is way off then IMO. It's somewhat similar but still rather different. The resulting behavior is an iterator yielding `Result` that promises to not yield more than one `Err`.

Reviewer 5: Which really fits with `Coroutine<(), Yield = T, Return = Result<!, E>>` better. It might make sense to not support `?` inside a `Coroutine<(), Yield = Result<T, E>, Return = ()>` based generator and instead require usage of `try` blocks.

Reviewer: This comparison stood out as unclear to me also, so we've now removed it from the document.

* ...and with an automatic `Iterator` implementation.

We'll probably be able to modularize the generator implementation and make it more robust on the implementation and diagnostics side for the `gen` block case, but I believe the initial implementation should be a HIR lowering to a generator and wrapping that generator in [`from_generator`][].

## Fusing

Special case the generator implementation if `Generator::Return = ()` to not panic, but
repeatedly produce values of the unit type.

# Drawbacks
[drawbacks]: #drawbacks

It's another language feature for something that can already be written entirely in user code.

In contrast to `Generator`, `gen` blocks that produce `Iterator`s cannot hold references across `yield` points.
See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.

[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives
## Keyword

We could use `iter` as the keyword.
I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.

> The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.

Reviewer 11: This probably needs updating now that that trait is getting renamed to `Coroutine`.

Also see [Wikipedia](https://en.wikipedia.org/wiki/Generator_(computer_programming)) -- it seems that "generator" is indeed usually tied to iterators.

Reviewer: The rename to `Coroutine` is now reflected throughout the document.

It may not be much worse than `gen` (see also [the unresolved questions][unresolved-questions]).
We may want to use `gen` for full on generators in the future.

## Do not do this

One alternative is to keep adding more helper methods to `Iterator`.
It is already hard for new Rustaceans to be aware of all the capabilities of `Iterator`.
Some of these new methods would need to be very generic.
While it's not an `Iterator` example, [`array::try_map`][] is something that has very complex diagnostics that are hard to improve, even if it's nice once it works.

Users can use crates like [`genawaiter`](https://crates.io/crates/genawaiter) or [`propane`](https://crates.io/crates/propane) instead.
`genawaiter` works on stable and provides `gen!` macro blocks that behave like `gen` blocks, but don't have compiler support for nice diagnostics or language support for the `?` operator. The `propane` crate uses the `Generator` trait from nightly and works mostly
like `gen` would.

The standard library includes [`std::iter::from_fn`][], which can be used in
some cases, but as we saw in the example [above][motivation], often the

> improvement over writing out a manual implementation of `Iterator` is limited.

Reviewer 11: Well, the example also shows that our existing combinators are better (more concise without being less clear) than generators. I think it'd be good to have an example where `gen` actually beats the status quo.

Reviewer 6: What would be the current way to [write this](https://github.com/estebank/iterator_item/blob/b79d7b496687b1a0f995bf16ca6482f37061dbca/tests/merge_overlapping_intervals.rs#L62-L73) that is more concise than the generator version?

Reviewer 11: I'm sure there's a lot of nice examples. :) I had to write iterator structs myself when I really just wanted `yield`. ([Recent example](https://github.com/rust-lang/rust/blob/f3f9b795bdaccd8284baba295810e87646754c28/compiler/rustc_const_eval/src/interpret/projection.rs#L99-L127).) I'm just saying the RFC should contain such an example.

Your code uses `?` so I'm already unsure what it will even do.^^ I think under this RFC it wouldn't work since `?` on `Option<T>` requires the yield type to be `Option<_>`?

Reviewer: We've now updated this example to implement run-length encoding.  This is much more *motivating*.

We could support using `await` in `gen async` blocks, similar to how we support `?` being used within `gen` blocks.
We'd have similar limitations holding references held across `await` points as we do have with `yield` points.
The solution space for `gen async` is large enough that I will not explore it here.
This RFC's design is forward compatible with anything we decide on.

At present it is only possible to have a `gen` block yield futures, but not `await` within it, similar to how
you cannot write iterators that `await`, but that return futures from `next`.

## Self-referential `gen` blocks

We can allow `gen` blocks to hold borrows across `yield` points in the future.

There are a few options forward (though this list is probably not complete):

* Add a separate trait for pinned iteration that is also usable with `gen` and `for`.
    * *Downside*: We would have very similar traits for the same thing.
* Backward-compatibly add a way to change the argument type of `Iterator::next`.
    * *Downside*: It's unclear whether this is possible.
* Implement `Iterator` for `Pin<&mut G>` instead of for `G` directly (whatever `G` is here, but it could be a `gen` block).

>     * *Downside*: The thing being iterated over must now be pinned for the entire iteration, instead of for each invocation of `next`.

Reviewer 11: I don't understand this point. Generally when something is pinned it has to stay pinned forever, you can't just unpin after one call to `next`,

Reviewer 13: I assume the point being made is that in the process of calling `iter.next()` in Rust today, the iterator *effectively* gets pinned in an earlier stack frame throughout `next`'s execution. It does seem like a strange way to conceptualise this though.

Reviewer: We've removed this point from the document.

- Feature Name: `gen-fn`
- Start Date: 2023-10-10
- RFC PR: [rust-lang/rfcs#3513](https://github.com/rust-lang/rfcs/pull/3513)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Reserve the `gen` keyword in the 2024 edition and start experimenting on
implementing generator functions and blocks that allow implementing `Iterator`s
without writing `Iterator::next` impls under placeholder syntax.

> Add `#[rustc_gen] {}` blocks to the language. These implement `Iterator` by `yield`ing

Reviewer: Thanks for this note.  We've removed the placeholder syntax.  This RFC will now commit to a specific syntax.

We're thinking more about what that syntax should be exactly.  We'll renominate this for T-lang when we're sure the answer is the best one.

I'll mark this as resolved as the placeholder syntax has been removed.

elements. This is simpler and more intuitive than creating a custom type and
manually implementing `Iterator` for that type, which requires writing an
explicit `Iterator::next` method body. This is a change similar to adding `async
{}` blocks that implement `Future` instead of having to manually write futures
and their state machines.

Furthermore, add `#[rustc_gen] fn` to the language. `#[rustc_gen] fn foo(arg: X) -> Y` desugars to
`fn foo(arg: X) -> impl Iterator<Item = Y>`.

# Motivation
[motivation]: #motivation

The main motivation of this RFC is to reserve a new keyword in the 2024 edition.

> The feature used by the keyword described here should be treated as an e-RFC for

Reviewer: Thanks for bringing this up.  We've removed the language about this being an experiment, so I'll mark this as resolved.

value becomes the result of the expression as usual.  When its argument
returns a value indicating that short-circuiting is desired
(e.g. `Option::None`, `Result::Err(..)`, `ControlFlow::Break(..)`), the value
is first yielded (after being converted by `From::from` as usual), then the
block returns immediately.

Even when `?` is used within a `#[rustc_gen]` block, the block must return a
value of type unit or `!`.  That is, it does not return a value of `Some(..)`,
`Ok(..)`, or `Continue(..)` as other such blocks might.

However, note that when `?` is used within a `#[rustc_gen]` block, all `yield`
statements will need to be given an argument of a compatible type.  For
example, if `None?` is used in an expression, then all `yield` statements will
need to be given arguments of type `Option`.

## Fusing

Like `Generators`, `Iterator`s produced by `#[rustc_gen]` panic when invoked again after they have returned `None` once.
This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially
produce infinite values of the unit type.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation
## New keyword

> In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `k#gen` to get the same features.

Reviewer: This is a good point and we've now addressed this in the RFC and left the handling for Rust 2015 and Rust 2018 as an open question.

        yield 1
    });

    assert_eq!(it.next(), Some(1));
    assert_eq!(it.next(), None);
    it.next(); // panics
}
```

## Panicking

What happens when `Iterator::next` is called again on a `#[rustc_gen]` block that panicked? Do we need to poison the iterator?

## Contextual keyword

Popular crates (like `rand`) have methods called [`gen`][Rng::gen]. If we forbid those, we are forcing those crates to make a major version bump when they update their edition, and we are requiring any users of those crates to use `r#gen` instead of `gen` when calling that method.

We could choose to use a contextual keyword and only forbid `gen` in:

* bindings
* field names (due to destructuring bindings)
* enum variants
* type names

This should avoid any parsing issues around `gen` followed by `{` in expressions.

[Rng::gen]: https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen

## `Iterator::size_hint`

Should we try to compute a conservative `size_hint`? This will reveal information from the body of a generator,
but at least for simple cases users will likely expect `size_hint` to not just be the default.

## Implement other `Iterator` traits.

> Is there a possibility for implementing traits like `DoubleEndedIterator`, `ExactSizeIterator` at all?

Reviewer 12: Musing: If we had fully-general RTN we could possibly allow implementing traits on them:

```rust
gen fn foo() -> i32 { ... }
impl ExactSizeIterator for foo() {}
```

We could also allow a single `impl Iterator for foo()` block to override `size_hint`.

It doesn't solve the problem of implementing `DoubleEndedIterator` with a generator though. That sounds more like we need a DoubleEndedGenerator with a `resume_back` method, with an attribute and compiler analysis to implement it.

Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings.
This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed

> or be referred to via `k#gen`.

Reviewer 10: No, the old version was correct. `r#` is for referring to identifiers that have since been shadowed by keywords, `k#` is for referring to keywords that don't yet exist.

Reviewer: Yes, oops, meant to change one line and  changed another.  Fixed both now.

## Fusing

`Iterator`s produced by `gen` keep returning `None` when invoked again after they have returned `None` once.
They do not implement `FusedIterator`, as that is not a language item, but may implement it in the future.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation
## New keyword

> In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `r#gen` to get the same features.

Reviewer 9: Why this change? I think `k#gen` was correct here for how to access the keyword in old editions (though we'll have to implement that). `r#gen` is what code in the new edition will use if they had something like a public API named `gen`.

Reviewer: That looks to have been a search/replace error in that commit.  I've pushed a fix now.
